content/asp.net/getting-started/data.md

```diff
@@ -1,20 +1,168 @@
+
+To see our data rendered in our browser at runtime we're going to need a [view](views.md). We'll follow [standard convention](controllers-actions.md) here, so add a `ViewErrors` folder under your `Views` directory in the project and add a new view called `Index.cshtml`. The requisite code to render a table of errors is as follows:
+
+```html
+@model IEnumerable<MvcBasics.Data.ErrorInformation>
+
+<table class="table">
+    @foreach (var error in Model)
+    {
+        <tr>
+            <td>@error.ErrorInformationId</td>
+            <td>@error.Path</td>
+            <td>@error.StatusCode</td>
+        </tr>
+    }
+</table>
+```
+
+When you restart your application you will have the ability to navigate to `/ViewErrors/` and see any that have surfaced. Remember that your in-memory database will have reset on each execution, so you may have to generate a few errors to get the data showing. Your end result should look something like this:
+
+![Viewing a List of Application Errors](images/error-table.png)
+
+## Filtering and Shaping Our Data
+
+The `Errors` property we have on our `ErrorContext` class is of type `DbSet<T>` which implements the `IEnumerable` and `IQueryable` interfaces. This enable many interesting Linq scenarios such as deferred execution, grouping and ordering. You can chain together various query parts before using the `ToList` extension method, and Entity Framework will take care of optimizing the query behind the scenes for you. For instance, if you wanted to extract a list of 404 errors by page and get a count of each, ordered by the path you could put together a Linq statement like so:
+
+```c#
+var notFoundErrors = _context.Errors
+    .Where(e => e.StatusCode == 404)
+    .GroupBy(e => e.Path)
+    .OrderBy(g => g.Key)
+    .ToList();
+```
+
+At runtime, the `IQueryable` components of the above statement defer execution until an iterator is invoked on the set. By putting off a query to the database until the data is needed, an optimized version of the query can be assembled and executed a single time, only returning the data that is required. For more information about working with data sets in Linq, be sure to run the the EF tutorial on the [docs site](https://docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/intro);
+
+
+## Next Steps
+ - Improve on this design by creating a view model class that can be used to store different sets of errors, grouped by type. Update the view to use this model, and pass it from the controller.
+ - Read more about the [repository pattern](https://msdn.microsoft.com/en-us/magazine/mt703433.aspx) and [mediator pattern](https://en.wikipedia.org/wiki/Mediator_pattern) and check out the [MediatR library](https://github.com/jbogard/MediatR) for an implementation. Use these patterns to remove the idea of the `DbContext` class from the controller.
```

Comment:
"Use these patterns to decouple the controller from the 'DbContext' class and the implementation details of how the app handles persistence."